File: safe-online-sdk/src/main/java/net/link/safeonline/sdk/ws/attrib/AttributeClient.java
/*
 * SafeOnline project.
 *
 * Copyright 2006-2007 Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

package net.link.safeonline.sdk.ws.attrib;

import java.io.Serializable;
import java.util.List;
import java.util.Map;
import net.link.safeonline.attribute.provider.AttributeSDK;
import net.link.safeonline.sdk.logging.exception.*;


/**
 * Interface for attribute client. Via components implementing this interface applications can retrieve attributes for subjects.
 * Applications can only retrieve attribute values for which the user confirmed the corresponding application identity.
 *
 * <p> The attribute value can be of type String, Boolean or an array of Object[] in case of a multivalued attribute. </p>
 *
 * @author fcorneli
 */
public interface AttributeClient {

    /**
     * Gives back attribute values via the map of attributes. The map should hold the requested attribute names as keys. The method will
     * fill in the corresponding values.
     *
     * @throws AttributeNotFoundException
     * @throws RequestDeniedException
     * @throws WSClientTransportException
     * @throws AttributeUnavailableException
     * @throws SubjectNotFoundException
     */
    void getAttributes(String userId, Map<String, List<AttributeSDK<?>>> attributes)
            throws AttributeNotFoundException, RequestDeniedException, WSClientTransportException, AttributeUnavailableException,
                   SubjectNotFoundException;

    /**
     * Gives back a map of attributes for the given subject that this application is allowed to read.
     *
     * @throws RequestDeniedException
     * @throws WSClientTransportException
     * @throws AttributeNotFoundException
     * @throws AttributeUnavailableException
     * @throws SubjectNotFoundException
     */
    Map<String, List<AttributeSDK<?>>> getAttributes(String userId)
            throws RequestDeniedException, WSClientTransportException, AttributeNotFoundException, AttributeUnavailableException,
                   SubjectNotFoundException;

    /**
     * Gives back a list of attributes for the given subject and attribute type that this application is allowed to read.
     *
     * @throws RequestDeniedException
     * @throws WSClientTransportException
     * @throws AttributeNotFoundException
     * @throws AttributeUnavailableException
     * @throws SubjectNotFoundException
     */
    <T extends Serializable> List<AttributeSDK<T>> getAttributes(String userId, String attributeName)
            throws RequestDeniedException, WSClientTransportException, AttributeNotFoundException, AttributeUnavailableException,
                   SubjectNotFoundException;
}

File: safe-online-sdk-dotnet/src/main/cs/ws/attrib/AttributeClient.cs
/*
 * SafeOnline project.
 * 
 * Copyright 2006-2008 	Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */
using System;
using System.Collections.Generic;

namespace safe_online_sdk_dotnet
{
	public interface AttributeClient
	{
		/// <summary>
		/// Gives back the attribute value of a single attribute of the given subject.
		/// This can be single-valued, multi-valued or compounded. 
		/// Compound attributes are returned with an array of Dictionaries.
		/// </summary>
		/// <param name="userId"></param>
		/// <param name="attributeName"></param>
		/// <returns></returns>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
		T getAttributeValue<T>(string userId, string attributeName);
		
		/// <summary>
		/// Gives back the attribute values via the specified dictionary of attributes. 
		/// The dictionary should hold the requested attribute names as keys. 
		/// The method will find the corresponding values.
		/// </summary>
		/// <param name="userId"></param>
		/// <param name="attributes"></param>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
		void getAttributeValues(string userId, Dictionary<string, object> attributes);
		
		/// <summary>
		/// Gives back a dictionary of attributes for the given subject that this application is allowed to read.
		/// </summary>
		/// <param name="userId"></param>
		/// <returns></returns>
		/// <exception cref="RuntimeException">Invalid response received</exception>
		/// <exception cref="AttributeNotFoundException">Attribute not found</exception>
		/// <exception cref="RequestDeniedException">Attribute Request denied</exception>
		/// <exception cref="AttributeUnavailableException">Attribute unavailable</exception>
		Dictionary<string, object> getAttributeValues(string userId);
	}
}
